### PR TITLE
Add cast platform

### LINKED ISSUE #233: [Feature Request] Spotify Casting capabilities in the media browser.
I want to make a feature to allow Spotify to cast to Chromecast devices from the media browser. 

I'm willing to do the brunt of the dev work on this.
I think the best place for a feature like this is in the official Spotify integration but as an unofficial Spotify API like the token you need to cast to a Chromecast device is unlikely to make it into home assistant core, I think this plugin is the best place to do the work.

I am unfamilia with the media browser API but as this plugin can call the official plugin I hope this as easy registering spotcast as a media source for Chromecast devices and then just using a shim and have official Spotify plugin to show the media browser. I realise it might be a lot more complicated then this but I think I would like to give it a try anyway.

Firstly I would like your blessing that you would accept a PR that does this.

Secondly I would like to use this thread to discuss implementation details.

### PR DESCRIPTION
**Pull request by emontnemery**
Add cast platform to spotcast, this enables cast media_player to play spotify content as well as browse media.

Media browsing relies entirely on the native Spotify integration.
Note that account information is not yet supported by the native Spotify integration's media browser.

This will be supported in Home Assistant Core 2022.2 pending merge of https://github.com/home-assistant/core/pull/65149
The cast platform is ignored by previous versions of Home Assistant Core, so the change is backwards compatible.

Closes #233 

Screenshots:
image
image
image

### CODE REVIEW

**Comment by fcusson** [comment]
retro compatibility and regression test were made successfully. Since we have time until the 2022.2 release, I will start to do a structure review. I don't think I saw any logging (errors or debug) which will have to be added before we merge the PR.

**Review by fcusson — CHANGES_REQUESTED** [review]
Changes to make

- [x] Correct specific changes requested in the document
- [x] Add Documentation regarding the functionality in README.md
- [ ] Add debug logs to indicate the process flow of the functionality

File: custom_components/spotcast/cast.py
Comment (emontnemery) [reply custom_components/spotcast/cast.py:9]: What will use that logger?

File: custom_components/spotcast/cast.py
Comment (emontnemery) [reply custom_components/spotcast/cast.py:30]: Is `from __future__ import annotations` OK instead?

File: custom_components/spotcast/cast.py
Comment (emontnemery) [reply custom_components/spotcast/cast.py]: Fixed in d19b846

File: custom_components/spotcast/cast.py
Comment (fcusson) [reply custom_components/spotcast/cast.py:30]: wasn't aware __futur__ was a thing. Of course you can use it.

File: custom_components/spotcast/cast.py
Comment (fcusson) [reply custom_components/spotcast/cast.py:9]: Yes, it is in reference to this comment.

I don't want us to be too complex in the debug logs we put. It's really for debugging when a user will eventually bring an issue regarding this workflow.

We could simply add a log stating that Spotcast is attempting X. Since we do get variables for `async_browse_media` and `async_play_media`, I would add this information in the log.

We should also check if we get scenarios where we run to the end of a function without returning properly something. If that's the case, we could check the behavior of Home assistant and check if an error log could provide more information.

Comment (emontnemery) [reply custom_components/spotcast/cast.py:9]: OK, got it 👍

File: custom_components/spotcast/cast.py
Comment (emontnemery) [reply custom_components/spotcast/cast.py:38]: Great, fixed in 5d8f4c7

File: custom_components/spotcast/cast.py
Comment (emontnemery) [reply custom_components/spotcast/cast.py:13]: Debug prints added in 5d8f4c7

**Comment by fcusson** [comment]
> @fcusson I think I covered all your comments, let me know if something else needs to be adjusted 👍 

I'll check after work

**Review by fcusson — APPROVED** [review]

**Comment by fondberg** [comment]
@emontnemery thanks! I will also have a look!

**Comment by emontnemery** [comment]
> it could be interesting to look into also adding a media_source.py

Spotify can't be a media source because it doesn't provide files or URLs.
The native integration does implement `async_browse_media` though to allow browsing the Spotify library.

What would you like to add to spotcast to extend or improve this further?

**Comment by fcusson** [comment]
> > it could be interesting to look into also adding a media_source.py
> 
> Spotify can't be a media source because it doesn't provide files or URLs. The native integration does implement `async_browse_media` though to allow browsing the Spotify library.
> 
> What would you like to add to spotcast to extend or improve this further?

I was just thinking of possible improvement without having checked if it was possible.

**Comment by fcusson** [comment]
We will have to do further testing, I just stress tested the functionality, and the Spotify token doesn't get refreshed which gave me 401 error while trying to browse the media.

I'm investigating if we are supposed to handle the token or if the Spotify integration should handle it.

**Comment by emontnemery** [comment]
> got token expired error after multiple attempt at using the media browser

What got a token expired error, spotcast or the native integration?

If the problem came from the native integration and it's setup to use account linking (the default) you may try setting it up to use a developer account instead. There's some trouble with the Home Assistant Cloud to Spotify account linking currently, it seems it's being throttled by Spotify.

**Comment by fcusson** [comment]
yeah, probably a throttling from Spotify (I already have the developper app setup) I've been stress testing it for a bit now and I wasn't able to get a 401 error again

**Review by fcusson — APPROVED** [review]
no longer getting 401 error. possibly due to throttling on Spotify part. I'm adding back my review

**Comment by emontnemery** [comment]
There will be some tweaks to the native Spotify integration in HA Core 2022.3 to allow browsing multiple Spotify accounts.
HA Core 2022.3 is scheduled for release in two weeks, so we may want to adapt this PR to the updated HA Core 2022.3 API and not merge it until HA Core 2022.3 is released.

**Comment by fcusson** [comment]
good point, does the documentation from the spotify integration points to how we can implement multiple accounts?

**Comment by emontnemery** [comment]
The relevant core changes are here: https://github.com/home-assistant/core/pull/66256, basically we should call `spotify.async_get_media_browser_root_object` to get the root nodes for browsing instead of hard coding.

spotcast itself does not need to worry about the multiple accounts as I understand it, since there's there doesn't seem to be a concept of private media in Spotify? Private user playlists can be played by any user who has a link t it?

**Comment by fcusson** [comment]
Since the code is working and we haven't got any feed back from other reviewers, I'll merge the code in the base. I have a fix I want to include in the next version, I'll add it with the media browser for the version.

**Pull request merged**
